File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/09．主动陷阱模板.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local jass = require("jass.common")
local japi = require("jass.japi")
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_0.addPeriodicCallback
local removePeriodicCallback = ____require_result_0.removePeriodicCallback
local getServerTime = ____require_result_0.getServerTime
local ____require_result_1 = require("lib.扩展函数.自定义扩展函数.01．选取中心范围")
local getUnitsInRange = ____require_result_1.getUnitsInRange
local ____require_result_2 = require("系统.03．技能系统.00．技能模板+函数.01．技能函数.16．扩展控制.扩展控制系统")
local _____65BD_52A0_6269_5C55_63A7_5236 = ____require_result_2["施加扩展控制"]
local AddSpecialEffect = jass.AddSpecialEffect
local DestroyEffect = jass.DestroyEffect
local GetOwningPlayer = jass.GetOwningPlayer
local IsUnitEnemy = jass.IsUnitEnemy
local IsUnitType = jass.IsUnitType
local GetUnitState = jass.GetUnitState
local UnitDamageTarget = jass.UnitDamageTarget
local UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD
local UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE
local ATTACK_TYPE_NORMAL = jass.ATTACK_TYPE_NORMAL
local DAMAGE_TYPE_MAGIC = jass.DAMAGE_TYPE_MAGIC
local WEAPON_TYPE_WHOKNOWS = jass.WEAPON_TYPE_WHOKNOWS
local EXSetEffectSize = japi.EXSetEffectSize
____exports["默认主动陷阱模型"] = "Abilities\\Spells\\Orc\\StasisTrap\\StasisTotemTarget.mdl"
local _____4E0B_4E00_4E2A_4E3B_52A8_9677_9631ID = 0
local function _____5355_4F4D_6709_6548_5B58_6D3B(unit)
    return unit ~= nil and unit ~= 0 and IsUnitType(unit, UNIT_TYPE_DEAD) ~= true and GetUnitState(unit, UNIT_STATE_LIFE) > 0.405
end
local function ____on_4E3B_52A8_9677_9631Tick(variable)
    local _____5B9E_4F8B = variable
    if _____5B9E_4F8B ~= nil then
        _____5B9E_4F8B["推进"](_____5B9E_4F8B)
    end
end
local _____4E3B_52A8_9677_9631_5B9E_73B0 = __TS__Class()
_____4E3B_52A8_9677_9631_5B9E_73B0.name = "主动陷阱实现"
function _____4E3B_52A8_9677_9631_5B9E_73B0.prototype.____constructor(self, ID, _____53C2_6570, _____7279_6548)
    self["回调ID"] = 0
    self["已销毁"] = false
    self.ID = ID
    self["名称"] = _____53C2_6570["名称"] or "主动陷阱#" .. tostring(ID)
    self["参数"] = _____53C2_6570
    self["特效"] = _____7279_6548
    self.X = _____53C2_6570.X
    self.Y = _____53C2_6570.Y
    self["到期时间"] = getServerTime() + _____53C2_6570["持续秒数"] * 1000
end
_____4E3B_52A8_9677_9631_5B9E_73B0.prototype["启动"] = function(self)
    if self["回调ID"] ~= 0 then
        return
    end
    self["回调ID"] = addPeriodicCallback(self["参数"]["检测间隔毫秒"] or 100, ____on_4E3B_52A8_9677_9631Tick, self)
end
_____4E3B_52A8_9677_9631_5B9E_73B0.prototype["销毁"] = function(self, _____539F_56E0)
    if _____539F_56E0 == nil then
        _____539F_56E0 = "手动销毁"
    end
    if self["已销毁"] then
        return
    end
    self["已销毁"] = true
    if self["回调ID"] ~= 0 then
        removePeriodicCallback(self["回调ID"])
        self["回调ID"] = 0
    end
    if self["特效"] ~= nil and self["特效"] ~= 0 then
        DestroyEffect(self["特效"])
    end
    if self["参数"]["on结束"] ~= nil then
        self["参数"]["on结束"](self, _____539F_56E0)
    end
end
_____4E3B_52A8_9677_9631_5B9E_73B0.prototype["推进"] = function(self)
    if self["已销毁"] then
        return
    end
    if getServerTime() >= self["到期时间"] then
        self["销毁"](self, "过期")
        return
    end
    local targets = getUnitsInRange(self.X, self.Y, self["参数"]["触发半径"])
    do
        local i = 0
        while i < #targets do
            do
                local target = targets[i + 1]
                if not self["目标可触发"](self, target) then
                    goto __continue17
                end
                self["触发"](self, target)
                return
            end
            ::__continue17::
            i = i + 1
        end
    end
end
_____4E3B_52A8_9677_9631_5B9E_73B0.prototype["目标可触发"] = function(self, target)
    if not _____5355_4F4D_6709_6548_5B58_6D3B(target) then
        return false
    end
    if self["参数"]["只触发敌人"] ~= false and not IsUnitEnemy(
        target,
        GetOwningPlayer(self["参数"]["施法者"])
    ) then
        return false
    end
    if self["参数"]["过滤目标"] ~= nil and not self["参数"]["过滤目标"](target, self) then
        return false
    end
    return true
end
_____4E3B_52A8_9677_9631_5B9E_73B0.prototype["触发"] = function(self, target)
    if self["参数"]["触发特效路径"] ~= nil and self["参数"]["触发特效路径"] ~= "" then
        local effect = AddSpecialEffect(self["参数"]["触发特效路径"], self.X, self.Y)
        if effect ~= nil and effect ~= 0 then
            DestroyEffect(effect)
        end
    end
    if (self["参数"]["触发伤害"] or 0) > 0 then
        local ____self__53C2_6570__653B_51FB_7C7B_578B_3 = self["参数"]["攻击类型"]
        if ____self__53C2_6570__653B_51FB_7C7B_578B_3 == nil then
            ____self__53C2_6570__653B_51FB_7C7B_578B_3 = ATTACK_TYPE_NORMAL
        end
        local attackType = ____self__53C2_6570__653B_51FB_7C7B_578B_3
        local ____self__53C2_6570__4F24_5BB3_7C7B_578B_4 = self["参数"]["伤害类型"]
        if ____self__53C2_6570__4F24_5BB3_7C7B_578B_4 == nil then
            ____self__53C2_6570__4F24_5BB3_7C7B_578B_4 = DAMAGE_TYPE_MAGIC
        end
        local damageType = ____self__53C2_6570__4F24_5BB3_7C7B_578B_4
        local ____self__53C2_6570__6B66_5668_7C7B_578B_5 = self["参数"]["武器类型"]
        if ____self__53C2_6570__6B66_5668_7C7B_578B_5 == nil then
            ____self__53C2_6570__6B66_5668_7C7B_578B_5 = WEAPON_TYPE_WHOKNOWS
        end
        local weaponType = ____self__53C2_6570__6B66_5668_7C7B_578B_5
        UnitDamageTarget(
            self["参数"]["施法者"],
            target,
            self["参数"]["触发伤害"] or 0,
            false,
            false,
            attackType,
            damageType,
            weaponType
        )
    end
    if self["参数"]["控制类型"] ~= nil and (self["参数"]["控制持续秒数"] or 0) > 0 then
        _____65BD_52A0_6269_5C55_63A7_5236(self["参数"]["施法者"], target, self["参数"]["控制类型"], {["持续时间"] = self["参数"]["控制持续秒数"] or 0})
    end
    if self["参数"]["on触发"] ~= nil then
        self["参数"]["on触发"](target, self)
    end
    if self["参数"]["触发后销毁"] ~= false then
        self["销毁"](self, "触发")
    end
end
____exports["创建主动陷阱"] = function(_____53C2_6570)
    if _____53C2_6570["施法者"] == nil or _____53C2_6570["施法者"] == 0 then
        return nil
    end
    if not (_____53C2_6570["持续秒数"] > 0) or not (_____53C2_6570["触发半径"] > 0) then
        return nil
    end
    local effect = AddSpecialEffect(_____53C2_6570["模型路径"] or ____exports["默认主动陷阱模型"], _____53C2_6570.X, _____53C2_6570.Y)
    if effect == nil or effect == 0 then
        return nil
    end
    if _____53C2_6570["缩放"] ~= nil and type(EXSetEffectSize) == "function" then
        EXSetEffectSize(effect, _____53C2_6570["缩放"])
    end
    local ____4E3B_52A8_9677_9631_5B9E_73B0_6 = _____4E3B_52A8_9677_9631_5B9E_73B0
    _____4E0B_4E00_4E2A_4E3B_52A8_9677_9631ID = _____4E0B_4E00_4E2A_4E3B_52A8_9677_9631ID + 1
    local _____5B9E_4F8B = __TS__New(____4E3B_52A8_9677_9631_5B9E_73B0_6, _____4E0B_4E00_4E2A_4E3B_52A8_9677_9631ID, _____53C2_6570, effect)
    _____5B9E_4F8B["启动"](_____5B9E_4F8B)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_7 = _____53C2_6570["清理"]
        ____self_7["登记清理"](
            ____self_7,
            _____5B9E_4F8B["名称"],
            function()
                _____5B9E_4F8B["销毁"](_____5B9E_4F8B)
            end
        )
    end
    return _____5B9E_4F8B
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/09．主动陷阱模板.ts
/** @noSelfInFile */

import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;

const { addPeriodicCallback, removePeriodicCallback, getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void, variable?: any) => void, variable?: any) => number;
  removePeriodicCallback: (this: void, id: number) => void;
  getServerTime: (this: void) => number;
};
const { getUnitsInRange } = require("lib.扩展函数.自定义扩展函数.01．选取中心范围") as {
  getUnitsInRange: (this: void, x: number, y: number, radius: number) => any[];
};
const { 施加扩展控制 } = require("系统.03．技能系统.00．技能模板+函数.01．技能函数.16．扩展控制.扩展控制系统") as {
  施加扩展控制: (this: void, 来源单位: any, 目标单位: any, 类型: 主动陷阱控制类型, 参数: { 持续时间: number }) => number;
};

const AddSpecialEffect = jass.AddSpecialEffect as (modelName: string, x: number, y: number) => any;
const DestroyEffect = jass.DestroyEffect as (effect: any) => void;
const GetOwningPlayer = jass.GetOwningPlayer as (unit: any) => any;
const IsUnitEnemy = jass.IsUnitEnemy as (whichUnit: any, whichPlayer: any) => boolean;
const IsUnitType = jass.IsUnitType as (unit: any, whichType: any) => boolean;
const GetUnitState = jass.GetUnitState as (unit: any, state: any) => number;
const UnitDamageTarget = jass.UnitDamageTarget as (
  whichUnit: any,
  target: any,
  amount: number,
  attack: boolean,
  ranged: boolean,
  attackType: any,
  damageType: any,
  weaponType: any
) => boolean;
const UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD as any;
const UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE as any;
const ATTACK_TYPE_NORMAL = jass.ATTACK_TYPE_NORMAL as any;
const DAMAGE_TYPE_MAGIC = jass.DAMAGE_TYPE_MAGIC as any;
const WEAPON_TYPE_WHOKNOWS = jass.WEAPON_TYPE_WHOKNOWS as any;
const EXSetEffectSize = japi.EXSetEffectSize as ((effect: any, size: number) => void) | undefined;

export const 默认主动陷阱模型 = "Abilities\\Spells\\Orc\\StasisTrap\\StasisTotemTarget.mdl";

export type 主动陷阱结束原因 = "触发" | "过期" | "手动销毁";
export type 主动陷阱控制类型 =
  | "stun"
  | "freeze"
  | "silence"
  | "polymorph"
  | "disarm"
  | "slow"
  | "stagger"
  | "pause"
  | "expause"
  | "sleep"
  | "roots"
  | "cyclone"
  | "taunt"
  | "charm"
  | "fear";

export interface 主动陷阱参数 {
  清理?: 机制清理篮子;
  名称?: string;
  施法者: any;
  X: number;
  Y: number;
  持续秒数: number;
  触发半径: number;
  检测间隔毫秒?: number;
  模型路径?: string;
  缩放?: number;
  触发后销毁?: boolean;
  只触发敌人?: boolean;
  过滤目标?: (this: void, 目标单位: any, 陷阱: 主动陷阱实例) => boolean;
  触发伤害?: number;
  伤害类型?: any;
  攻击类型?: any;
  武器类型?: any;
  控制类型?: 主动陷阱控制类型;
  控制持续秒数?: number;
  触发特效路径?: string;
  on触发?: (this: void, 目标单位: any, 陷阱: 主动陷阱实例) => void;
  on结束?: (this: void, 陷阱: 主动陷阱实例, 原因: 主动陷阱结束原因) => void;
}

export interface 主动陷阱实例 {
  readonly ID: number;
  readonly 名称: string;
  readonly 特效: any;
  readonly X: number;
  readonly Y: number;
  销毁(原因?: 主动陷阱结束原因): void;
}

let 下一个主动陷阱ID = 0;

function 单位有效存活(this: void, unit: any): boolean {
  return unit != null && unit !== 0 && IsUnitType(unit, UNIT_TYPE_DEAD) !== true && GetUnitState(unit, UNIT_STATE_LIFE) > 0.405;
}

function on主动陷阱Tick(this: void, variable?: any): void {
  const 实例 = variable as 主动陷阱实现 | undefined;
  if (实例 != null) 实例.推进();
}

class 主动陷阱实现 implements 主动陷阱实例 {
  readonly ID: number;
  readonly 名称: string;
  readonly 特效: any;
  readonly X: number;
  readonly Y: number;
  private 参数: 主动陷阱参数;
  private 到期时间: number;
  private 回调ID = 0;
  private 已销毁 = false;

  constructor(ID: number, 参数: 主动陷阱参数, 特效: any) {
    this.ID = ID;
    this.名称 = 参数.名称 ?? ("主动陷阱#" + String(ID));
    this.参数 = 参数;
    this.特效 = 特效;
    this.X = 参数.X;
    this.Y = 参数.Y;
    this.到期时间 = getServerTime() + 参数.持续秒数 * 1000;
  }

  启动(): void {
    if (this.回调ID !== 0) return;
    this.回调ID = addPeriodicCallback(this.参数.检测间隔毫秒 ?? 100, on主动陷阱Tick, this);
  }

  销毁(原因: 主动陷阱结束原因 = "手动销毁"): void {
    if (this.已销毁) return;
    this.已销毁 = true;
    if (this.回调ID !== 0) {
      removePeriodicCallback(this.回调ID);
      this.回调ID = 0;
    }
    if (this.特效 != null && this.特效 !== 0) DestroyEffect(this.特效);
    if (this.参数.on结束 != null) this.参数.on结束(this, 原因);
  }

  推进(): void {
    if (this.已销毁) return;
    if (getServerTime() >= this.到期时间) {
      this.销毁("过期");
      return;
    }
    const targets = getUnitsInRange(this.X, this.Y, this.参数.触发半径);
    for (let i = 0; i < targets.length; i++) {
      const target = targets[i];
      if (!this.目标可触发(target)) continue;
      this.触发(target);
      return;
    }
  }

  private 目标可触发(target: any): boolean {
    if (!单位有效存活(target)) return false;
    if (this.参数.只触发敌人 !== false && !IsUnitEnemy(target, GetOwningPlayer(this.参数.施法者))) return false;
    if (this.参数.过滤目标 != null && !this.参数.过滤目标(target, this)) return false;
    return true;
  }

  private 触发(target: any): void {
    if (this.参数.触发特效路径 != null && this.参数.触发特效路径 !== "") {
      const effect = AddSpecialEffect(this.参数.触发特效路径, this.X, this.Y);
      if (effect != null && effect !== 0) DestroyEffect(effect);
    }
    if ((this.参数.触发伤害 ?? 0) > 0) {
      const attackType = this.参数.攻击类型 ?? ATTACK_TYPE_NORMAL;
      const damageType = this.参数.伤害类型 ?? DAMAGE_TYPE_MAGIC;
      const weaponType = this.参数.武器类型 ?? WEAPON_TYPE_WHOKNOWS;
      UnitDamageTarget(
        this.参数.施法者,
        target,
        this.参数.触发伤害 ?? 0,
        false,
        false,
        attackType,
        damageType,
        weaponType
      );
    }
    if (this.参数.控制类型 != null && (this.参数.控制持续秒数 ?? 0) > 0) {
      施加扩展控制(this.参数.施法者, target, this.参数.控制类型, { 持续时间: this.参数.控制持续秒数 ?? 0 });
    }
    if (this.参数.on触发 != null) this.参数.on触发(target, this);
    if (this.参数.触发后销毁 !== false) this.销毁("触发");
  }
}

export function 创建主动陷阱(this: void, 参数: 主动陷阱参数): 主动陷阱实例 | null {
  if (参数.施法者 == null || 参数.施法者 === 0) return null;
  if (!(参数.持续秒数 > 0) || !(参数.触发半径 > 0)) return null;
  const effect = AddSpecialEffect(参数.模型路径 ?? 默认主动陷阱模型, 参数.X, 参数.Y);
  if (effect == null || effect === 0) return null;
  if (参数.缩放 != null && typeof EXSetEffectSize === "function") EXSetEffectSize(effect, 参数.缩放);

  const 实例 = new 主动陷阱实现(++下一个主动陷阱ID, 参数, effect);
  实例.启动();
  if (参数.清理 != null) {
    参数.清理.登记清理(实例.名称, function 主动陷阱清理(this: void): void {
      实例.销毁();
    });
  }
  return 实例;
}

export {};
